File: java/ReadWrite.java
/*
 * @Author:	Jeff Berube
 * @Title:	ReadWrite (Java)
 * @Description:	Implements a solution to the reader writer problem with reader
 * 			bias using semaphores.
 *
 * @Usage:	java ReadWrite
 *
 */

import java.util.concurrent.Semaphore;

public class ReadWrite {

	/* Define class variables */
	private static final int NUM_READERS = 4;
	private static final int NUM_WRITERS = 1;

	private static final Semaphore x = new Semaphore(NUM_READERS, true);
	private static final Semaphore wsem = new Semaphore(NUM_WRITERS, true);
	private static int readcount = 0;

	private static Thread[] threads = new Thread[NUM_READERS + NUM_WRITERS];

	public static void main(String[] args) {

		System.out.println("ReadWrite started with " + NUM_READERS +
					" readers and " + NUM_WRITERS + " writers.");

		int i = 0;

		/* Create reader threads */
		while (i < NUM_READERS) {

			threads[i] = new Thread(new Reader(i + 1));
			threads[i].start();
	
			i++;

		}

		i = 0;

		/* Create writer threads */
		while (i < NUM_WRITERS) {
		
			threads[i + NUM_READERS] = new Thread(new Writer(i + 1));
			threads[i + NUM_READERS].start();
			
			i++;

		}


		/* Wait on threads to finish */
		i = 0;

		while (i < (NUM_READERS + NUM_WRITERS)) {
		
			try {

				threads[i++].join();
			
			} catch (InterruptedException ie) {}

		}

		/* One last little output for prettiness */
		System.out.println("\n\n");

	}	

	static class Reader implements Runnable {

		private int threadnum;

		public Reader(int threadnum) {

			this.threadnum = threadnum;

		}

		public void run() {

			while (true) {
		
				/* Sleep */
				try {			
				
					Thread.sleep(threadnum * 3000);
					x.acquire();


					/* Acquire semaphore and increase reader count */
					readcount++;

					/* If first reader, acquire writer semaphore */
					if (readcount == 1) wsem.acquire();

					/* Release reader count semaphore */
					x.release();
			
					/* Display thread status, sleep and 
					 * display status again */
					System.out.println("Reader " + threadnum
						+ ": Inside critical section. Reading.");

					Thread.sleep(threadnum * 1000);
				
					System.out.println("Reader " + threadnum
						+ ": Done reading. Exiting critical section.");

					/* Down on reader count semaphore and 
					 * update reader count */
					x.acquire();
					readcount--;

					/* If we were last reader, release writer semaphore */
					if (readcount == 0) wsem.release();

					/* Release reader count semaphore */
					x.release();
				
				} catch (InterruptedException e) {}

			}
		
		}

	}

	
	static class Writer implements Runnable {

		private int threadnum;

		public Writer(int threadnum) {

			this.threadnum = threadnum;

		}

		public void run() {

			while (true) {
				
				try {

					/* Sleep */
					Thread.sleep(threadnum * 3000);
			
					/* Acuire writer semaphore */
					wsem.acquire();

					/* Display status, sleep, update status */
					System.out.println("Writer " + threadnum
						+ ": Inside critical section. Writing.");

					Thread.sleep(threadnum * 1000);
			
					System.out.println("Writer " + threadnum
						+ ": Done writing. Exiting critical section.");

					/* Release writer semaphore */
					wsem.release();
			
				} catch (InterruptedException e) {}

			}

		}

	}

}

File: c/readwrite.c
/*
 * @Author:	Jeff Berube
 * @Title:	Reader Writer	
 *
 * @Usage	./readwrite
 *
 * @Description:	Implements a solution to the reader writer problem
 * 			with reader bias using semaphores.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <time.h>
#include <unistd.h>

#define NUM_READERS	4
#define NUM_WRITERS	1

int readcount = 0;
sem_t x, wsem;

/*
 * Reader 
 * 
 * Thread function. Simulates a reader.
 *
 */

void *reader(void *param) {

	/* Initialize variables */
	int threadnum = (int)param;

	/* Seed random number generator */
	srand(time(NULL));

	while (1) {

		/* Sleep before trying to read */
		usleep(threadnum * 3000000);

		/* Down on reader count semaphore and update reader count */
		sem_wait(&x);
		readcount++;

		/* If we're first reader, lock writing semaphore */
		if (readcount == 1) sem_wait(&wsem);

		/* Release reader count semaphore so other readers can come in */
		sem_post(&x);

		/* Display status, sleep and display status again */
		printf("Reader %d: Inside critical section. Reading.\n", threadnum);
		usleep(threadnum * 1500000);
		printf("Reader %d: Done reading. Exiting critical section.\n", threadnum);

		/* Down on reader count semaphore and update reader count */
		sem_wait(&x);
		readcount--;

		/* If we were last reader and there is no readers after us, release
		 * writer semaphore
		 */
		if (readcount == 0) sem_post(&wsem);

		/* Release reader count semaphore */
		sem_post(&x);

	}

	return NULL;

}

/*
 * Writer
 *
 * Thread function. Simulates a write.
 *
 */

void *writer(void *param) {

	/* Variables */
	int threadnum = (int)param;

	/* Seed random number generator */
	srand(time(NULL));

	while (1) {
		
		/* Sleep before writing again */
		usleep(threadnum * 3000000);

		/* Acquire writer semaphore */
		sem_wait(&wsem);

		/* Display status, sleep and display status again */
		printf("Writer %d: Inside critical section. Writing.\n", threadnum);
		usleep(threadnum * 1500000);
		printf("Writer %d: Done writing. Exiting critical section.\n", threadnum);

		/* Release writer semaphore */
		sem_post(&wsem);

	}

	return NULL;
}

int main (int argc, char *argv[]) {

	printf("readwrite started with %d readers and %d writers.\n", NUM_READERS,
			NUM_WRITERS);

	/* Create reader and writer threads */
	pthread_t *threads = malloc((NUM_READERS + NUM_WRITERS) * sizeof(pthread_t));

	/* Thread variables */
	pthread_t tid;
	pthread_attr_t attr;

	/* Init semaphores */
	sem_init(&x, 0, 1);
	sem_init(&wsem, 0, 1);

	/* Initiate and create thread */
	pthread_attr_init(&attr);

	int i = 0;

	/* Create reader threads */
	while (i < NUM_READERS) {
		
		pthread_create(threads + i, &attr, reader, (void *)i + 1);
		i++;

	}

	i = 0;
	
	/* Create writer threads */
	while (i < NUM_WRITERS) {
	
		pthread_create(threads + i + NUM_READERS, &attr, writer, (void *)i + 1);
		i++;

	}

	i = 0;

	/* Wait on threads to finish. Never happens but program exits if we don't. */
	while (i < (NUM_WRITERS + NUM_READERS)) {
	
		pthread_join(*(threads + i), NULL);
		i++;

	}

	return  0;

}
